### DISCUSSION TITLE
How to evaluate execution units before user signature

### DISCUSSION THREAD

**Discussion by MartinSchere** [CardanoSolutions/ogmios#184] [Q & A] [answered]
Hello @KtorZ,

Appreciate all the awesome work put into this library!

I have a question: We want to use the exunits calculator but we are struggling to find a way to make the validator run without errors, since it requires signatures from several pubkey hashes. And if we have the user sign the transaction and then change the redeemer, the `script_data_hash` is a mismatch thus invalidating the whole transaction.

Any ideas on how to calculate them properly?

**Comment by KtorZ** [comment]
If I understand your case right you have a transaction with some validator which verifies the presence of some signatures on the transaction body. So you can probably "just": 

1. Construct a first transaction template, with evaluation units of redeemers set to 0
2. Sign this transaction
3. Evaluate the execution units
4. Replace the execution units with their correct value, re-calculate the script integrity hash and re-sign the transaction. 

Note that, doing (2) may be delicate depending on your scenario; if the signature is interactive here, then you'd basically be asking for the signature twice to the user, which is quite a bad UX overall. If you are somewhat in control of the keys, then it's no big deal. 

Also, depending on your scenario, it is maybe possible to sign the transaction with _another key_? In the end, what you care about really is the execution units of a transaction _of that kind_; and it doesn't matter that the transaction is made with Alice's key; Bob's key would cost the same. And I really meant, for two analogous transactions in two separate contexts, so not saying that two different key would validate the same transaction.. but that you really have one prototypical transaction template, and that regardless of the configuration / parameters you use, it will have one price. So, you could determine the execution units of such a prototypical transaction once; and then, use it for ALL subsequent transactions in your app. Thus, you wouldn't even need to do this dynamically as part of your transaction flow.

**Reply by MartinSchere** [reply]
Thank you for your answer.

The two-signatures approach is not an option for us at the moment.

I had thought about this "mocking" strategy, however, different users have different UTxOs which affect the resources that a script consumes. For example, a user might have 15 big UTxOs and multiple outputs they are using for the current transaction, which the script needs to validate in order to check for specific conditions. So there would be no way for us to do this dynamically and have each user pay exactly for the execution units they are using.

**Comment by MartinSchere** [comment, marked as answer]
@KtorZ Had a conversation with the great Alessandro, we just needed to set the required signers beforehand, since this phase of evaluation assumes that the signatories are already valid. All credits to him for finding this out.
